sovryn-token-bridge/federator/src/lib/Federator.js

```diff
@@ -157,55 +157,72 @@ module.exports = class Federator {
+                log.transactionHash,
+                log.logIndex,
+                decimals,
+                granularity,
+                userData || []
+            ).call();
+
+            this.logger.info('get transaction id U:', transactionIdU);
+            let wasProcessedU = await this.federationContract.methods.transactionWasProcessed(transactionIdU).call();
+            this.logger.info('wasProcessedU:', wasProcessedU);
+            if (!wasProcessedU) {
+                let hasVoted = await this.federationContract.methods.hasVoted(transactionIdU).call({ from: from });
+                if (!hasVoted) {
+                    this.logger.info(`Voting tx: ${log.transactionHash} block: ${log.blockHash} token: ${symbol}`);
+                    await this._voteTransaction(tokenAddress,
+                        receiver,
+                        amount,
+                        symbol,
+                        log.blockHash,
+                        log.transactionHash,
+                        log.logIndex,
+                        decimals,
+                        granularity,
+                        userData,
+                        transactionId,
+                        transactionIdU
+                    );
+                } else {
+                    this.logger.debug(`Block: ${log.blockHash} Tx: ${log.transactionHash} token: ${symbol}  has already been voted by us`);
+                }
             }
         } else {
             this.logger.debug(`Block: ${log.blockHash} Tx: ${log.transactionHash} token: ${symbol} was already processed`);
         }
     }
 
 
-    async _voteTransaction(tokenAddress, receiver, amount, symbol, blockHash, transactionHash, logIndex, decimals, granularity, userData) {
-        let txId;
+    async _voteTransaction(tokenAddress, receiver, amount, symbol, blockHash, transactionHash, logIndex, decimals, granularity, userData, transactionId, transactionIdU) {
         try {
 
             const transactionSender = new TransactionSender(this.sideWeb3, this.logger, this.config);
             this.logger.info(`Voting Transfer ${amount} of ${symbol} trough sidechain bridge ${this.sideBridgeContract.options.address} to receiver ${receiver}`);
 
-            txId = await this.federationContract.methods.getTransactionId(
-                tokenAddress,
-                receiver,
-                amount,
-                symbol,
-                blockHash,
-                transactionHash,
-                logIndex,
-                decimals,
-                granularity
-            ).call();
-
-            if (this.failingTxIds.contains(txId)) {
-                this.logger.info(
-                    `Transaction with id ${txId}, hash: ${transactionHash} is marked as failing -- not voting for it`
-                );
-                return false;
-            }
+            const isFailing = this._isFailingByTxId(transactionId, transactionIdU, transactionHash)
+            if (!isFailing) return false
```

Comment:
This is confusing because `_isFailingByTxId` returns `false` if the transaction is failing and `true` if it's not failing, ie. exectly the opposite. The logic works though.

Reply:
@koirikivi  you are right it is confusing
changed it